File: ios/Runner/SekretessCryptographicService.swift
import Foundation
import LibSignalClient

/// Main Signal Protocol cryptographic service for iOS.
/// Mirrors the functionality of the Android SekretessCryptographicService.
class SekretessCryptographicService {
    private let SIGNAL_KEY_COUNT = 15
    private let signalProtocolStore: SekretessSignalProtocolStore
    private let deviceId: UInt32 = 1
    private let storeContext = SimpleStoreContext()
    
    init(signalProtocolStore: SekretessSignalProtocolStore) {
        self.signalProtocolStore = signalProtocolStore
    }
    
    /// Initialize Signal Protocol - generates keys if needed and uploads to server
    func initialize() throws -> Bool {
        if signalProtocolStore.registrationRequired() {
            let keyBundle = try initializeKeyBundle()
            
            // Call API bridge to upload keys
            let keyBundleMap = KeyBundleConverter.toMap(keyBundle: keyBundle)
            let success = FlutterDependencyProvider.callApi(
                method: "upsertKeyStore",
                arguments: keyBundleMap
            )
            
            if success {
                storeKyberPreKeyRecords(keyBundle.kyberPreKeyRecords)
                storePreKeyRecords(keyBundle.opk)
                storeSignedPreKey(keyBundle.signedPreKeyRecord)
                return true
            } else {
                print("SekretessCryptographicService: Upsert cryptographic keys failed")
                signalProtocolStore.clearStorage()
                return false
            }
        } else if signalProtocolStore.updateKeysRequired() {
            print("SekretessCryptographicService: Update onetime cryptographic keys")
            try updateOneTimeKeys()
        }
        return true
    }
    
    /// Decrypt a group chat message
    func decryptGroupChatMessage(sender: String, base64Message: String) throws -> String? {
        guard let messageData = Data(base64Encoded: base64Message) else {
            throw SignalProtocolError.invalidMessage
        }
        
        let address = try ProtocolAddress(name: sender, deviceId: deviceId)
        
        do {
            let decryptedData = try groupDecrypt(
                messageData,
                from: address,
                store: signalProtocolStore.senderKeyStore,
                context: storeContext
            )
            guard let decryptedString = String(data: decryptedData, encoding: .utf8) else {
                throw SignalProtocolError.invalidMessage
            }
            return decryptedString
        } catch {
            print("Error decrypting group chat message: \(error)")
            throw SignalProtocolError.invalidMessage
        }
    }
    
    /// Decrypt a private message
    func decryptPrivateMessage(sender: String, base64Message: String) throws -> String? {
        guard let messageData = Data(base64Encoded: base64Message) else {
            throw SignalProtocolError.invalidMessage
        }
        
        let address = try ProtocolAddress(name: sender, deviceId: deviceId)
        
        // Try as PreKeySignalMessage first
        do {
            let preKeyMessage = try PreKeySignalMessage(bytes: messageData)
            let decryptedData = try signalDecryptPreKey(
                message: preKeyMessage,
                from: address,
                sessionStore: signalProtocolStore.sessionStore,
                identityStore: signalProtocolStore.identityKeyStore,
                preKeyStore: signalProtocolStore.preKeyStore,
                signedPreKeyStore: signalProtocolStore.signedPreKeyStore,
                kyberPreKeyStore: signalProtocolStore.kyberPreKeyStore,
                context: storeContext,
                usePqRatchet: true
            )
            guard let decryptedString = String(data: decryptedData, encoding: .utf8) else {
                throw SignalProtocolError.invalidMessage
            }
            return decryptedString
        } catch {
            // Try as SignalMessage
            do {
                let signalMessage = try SignalMessage(bytes: messageData)
                let decryptedData = try signalDecrypt(
                    message: signalMessage,
                    from: address,
                    sessionStore: signalProtocolStore.sessionStore,
                    identityStore: signalProtocolStore.identityKeyStore,
                    context: storeContext
                )
                guard let decryptedString = String(data: decryptedData, encoding: .utf8) else {
                    throw SignalProtocolError.invalidMessage
                }
                return decryptedString
            } catch {
                print("Error decrypting private message: \(error)")
                throw error
            }
        }
    }
    
    /// Process a key distribution message for group chats
    func processKeyDistributionMessage(name: String, base64Key: String) throws {
        guard let keyData = Data(base64Encoded: base64Key) else {
            throw SignalProtocolError.invalidMessage
        }
        
        let address = try ProtocolAddress(name: name, deviceId: deviceId)
        let distributionMessage = try SenderKeyDistributionMessage(bytes: keyData)
        
        do {
            try processSenderKeyDistributionMessage(
                distributionMessage,
                from: address,
                store: signalProtocolStore.senderKeyStore,
                context: storeContext
            )
            print("SekretessCryptographicService: Group chat cipher created and stored: \(name)")
        } catch {
            print("Error processing key distribution message: \(error)")
            throw error
        }
    }
    
    /// Update one-time keys
    func updateOneTimeKeys() throws {
        let identityKeyPair = try signalProtocolStore.identityKeyStore.identityKeyPair(context: storeContext)
        let preKeyRecords = generatePreKeys()
        let kyberPreKeyRecords = generateKyberPreKeys(privateKey: identityKeyPair.privateKey)
        
        // Convert to map for API call
        let keysMap = KeyBundleConverter.oneTimeKeysToMap(
            preKeyRecords: preKeyRecords,
            kyberPreKeyRecords: kyberPreKeyRecords
        )
        
        let success = FlutterDependencyProvider.callApi(
            method: "updateOneTimeKeys",
            arguments: keysMap
        )
        
        if success {
            storePreKeyRecords(preKeyRecords)
            storeKyberPreKeyRecords(kyberPreKeyRecords)
        } else {
            throw SignalProtocolError.apiCallFailed
        }
    }
    
    /// Initialize key bundle for signup
    func initializeKeyBundle() throws -> KeyBundle {
        signalProtocolStore.clearStorage()
        
        let signedPreKeyPrivateKey = PrivateKey.generate()
        let identityKeyPair = try signalProtocolStore.identityKeyStore.identityKeyPair(context: storeContext)
        let registrationId = try signalProtocolStore.identityKeyStore.localRegistrationId(context: storeContext)
        
        let publicKeyData = signedPreKeyPrivateKey.publicKey.serialize()
        let signature = identityKeyPair.privateKey.generateSignature(message: publicKeyData)
        
        let opk = generatePreKeys()
        let signedPreKeyRecord = try generateSignedPreKey(privateKey: signedPreKeyPrivateKey, signature: signature)
        let kyberPreKeyRecords = generateKyberPreKeys(privateKey: identityKeyPair.privateKey)
        
        return KeyBundle(
            registrationId: registrationId,
            opk: opk,
            signedPreKeyRecord: signedPreKeyRecord,
            identityKeyPair: identityKeyPair,
            signature: signature,
            kyberPreKeyRecords: kyberPreKeyRecords
        )
    }
    
    private func generatePreKeys() -> [PreKeyRecord] {
        var preKeys: [PreKeyRecord] = []
        for _ in 0..<SIGNAL_KEY_COUNT {
            let id = UInt32.random(in: 1..<9999999)
            let privateKey = PrivateKey.generate()
            do {
                let preKey = try PreKeyRecord(id: id, privateKey: privateKey)
                preKeys.append(preKey)
            } catch {
                print("Error generating pre key: \(error)")
                // Continue with next key
            }
        }
        return preKeys
    }
    
    private func generateKyberPreKeys(privateKey: PrivateKey) -> [KyberPreKeyRecord] {
        var kyberPreKeys: [KyberPreKeyRecord] = []
        // Generate SIGNAL_KEY_COUNT + 1 (last resort key)
        for _ in 0..<(SIGNAL_KEY_COUNT + 1) {
            let id = UInt32.random(in: 1..<9999999)
            let kemKeyPair = KEMKeyPair.generate()
            do {
                let publicKeyData = kemKeyPair.publicKey.serialize()
                let signature = privateKey.generateSignature(message: publicKeyData)
                let timestamp = UInt64(Date().timeIntervalSince1970)
                let kyberPreKey = try KyberPreKeyRecord(
                    id: id,
                    timestamp: timestamp,
                    keyPair: kemKeyPair,
                    signature: signature
                )
                kyberPreKeys.append(kyberPreKey)
            } catch {
                print("Error generating kyber pre key: \(error)")
                // Continue with next key
            }
        }
        return kyberPreKeys
    }
    
    private func generateSignedPreKey(privateKey: PrivateKey, signature: Data) throws -> SignedPreKeyRecord {
        let id = UInt32.random(in: 1..<9999999)
        let timestamp = UInt64(Date().timeIntervalSince1970)
        return try SignedPreKeyRecord(
            id: id,
            timestamp: timestamp,
            privateKey: privateKey,
            signature: signature
        )
    }
    
    private func storePreKeyRecords(_ records: [PreKeyRecord]) {
        for record in records {
            do {
                try signalProtocolStore.preKeyStore.storePreKey(record, id: record.id, context: storeContext)
            } catch {
                print("Error storing pre key: \(error)")
            }
        }
    }
    
    private func storeKyberPreKeyRecords(_ records: [KyberPreKeyRecord]) {
        for record in records {
            do {
                try signalProtocolStore.kyberPreKeyStore.storeKyberPreKey(record, id: record.id, context: storeContext)
            } catch {
                print("Error storing kyber pre key: \(error)")
            }
        }
    }
    
    private func storeSignedPreKey(_ record: SignedPreKeyRecord) {
        do {
            try signalProtocolStore.signedPreKeyStore.storeSignedPreKey(record, id: record.id, context: storeContext)
        } catch {
            print("Error storing signed pre key: \(error)")
        }
    }
}

enum SignalProtocolError: Error {
    case invalidMessage
    case apiCallFailed
    case initializationFailed
    case invalidKeyId
    case noSession
    case duplicateMessage
    case legacyMessage
    case invalidVersion
    case invalidKey
    case untrustedIdentity
}

File: android/app/src/main/java/io/sekretess/service/SekretessCryptographicService.java
package io.sekretess.service;

import android.util.Log;
import android.widget.Toast;

import org.signal.libsignal.protocol.DuplicateMessageException;
import org.signal.libsignal.protocol.IdentityKeyPair;
import org.signal.libsignal.protocol.InvalidKeyException;
import org.signal.libsignal.protocol.InvalidKeyIdException;
import org.signal.libsignal.protocol.InvalidMessageException;
import org.signal.libsignal.protocol.InvalidVersionException;
import org.signal.libsignal.protocol.LegacyMessageException;
import org.signal.libsignal.protocol.NoSessionException;
import org.signal.libsignal.protocol.SessionCipher;
import org.signal.libsignal.protocol.SignalProtocolAddress;
import org.signal.libsignal.protocol.UntrustedIdentityException;
import org.signal.libsignal.protocol.UsePqRatchet;
import org.signal.libsignal.protocol.ecc.ECKeyPair;
import org.signal.libsignal.protocol.ecc.ECPrivateKey;
import org.signal.libsignal.protocol.groups.GroupCipher;
import org.signal.libsignal.protocol.groups.GroupSessionBuilder;
import org.signal.libsignal.protocol.kem.KEMKeyPair;
import org.signal.libsignal.protocol.kem.KEMKeyType;
import org.signal.libsignal.protocol.message.PreKeySignalMessage;
import org.signal.libsignal.protocol.message.SenderKeyDistributionMessage;
import org.signal.libsignal.protocol.state.KyberPreKeyRecord;
import org.signal.libsignal.protocol.state.PreKeyRecord;
import org.signal.libsignal.protocol.state.SignedPreKeyRecord;
import org.signal.libsignal.protocol.util.Medium;

import java.security.SecureRandom;
import java.util.Base64;
import java.util.Optional;

import io.sekretess.bridge.FlutterDependencyProvider;
import io.sekretess.bridge.NativeApiClientBridge;
import io.sekretess.dto.KeyBundle;
import io.sekretess.cryptography.storage.SekretessSignalProtocolStore;


public class SekretessCryptographicService {
    private final int SIGNAL_KEY_COUNT = 15;
    private static final Base64.Decoder base64Decoder = Base64.getDecoder();
    private final SekretessSignalProtocolStore sekretessSignalProtocolStore;
    private final String TAG = "SekretessCryptographicService";
    private final int deviceId = 1;
    private NativeApiClientBridge apiClientBridge;


    public SekretessCryptographicService(SekretessSignalProtocolStore sekretessSignalProtocolStore) {
        this.sekretessSignalProtocolStore = sekretessSignalProtocolStore;
    }
    
    public void setApiClientBridge(NativeApiClientBridge apiClientBridge) {
        this.apiClientBridge = apiClientBridge;
    }
    
    public SekretessSignalProtocolStore getSignalProtocolStore() {
        return sekretessSignalProtocolStore;
    }

    public void updateOneTimeKeys() {
        IdentityKeyPair identityKeyPair = sekretessSignalProtocolStore.getIdentityKeyPair();
        PreKeyRecord[] preKeyRecords = generatePreKeys();
        KyberPreKeyRecord[] kyberPreKeyRecords = generateKyberPreKeys(identityKeyPair.getPrivateKey());
        try {
            if (apiClientBridge != null && apiClientBridge.updateOneTimeKeys(preKeyRecords, kyberPreKeyRecords)) {
                storePreKeyRecords(preKeyRecords);
                storeKyberPreKeyRecords(kyberPreKeyRecords);
            }
        } catch (Exception e) {
            Log.e(TAG, "Error during update one time keys", e);
            if (apiClientBridge != null) {
                apiClientBridge.showToast("Error during update one time keys: " + e.getMessage());
            }
        }
    }

    public KyberPreKeyRecord[] generateKyberPreKeys(ECPrivateKey ecPrivateKey) {
        // Generate post quantum resistance keys + 1 last resort key
        KyberPreKeyRecord[] kyberPreKeyRecords = new KyberPreKeyRecord[SIGNAL_KEY_COUNT + 1];

        for (int i = 0; i < kyberPreKeyRecords.length; i++) {
            KyberPreKeyRecord kyberPreKeyRecord = generateKyberPreKey(ecPrivateKey);
            kyberPreKeyRecords[i] = kyberPreKeyRecord;
        }
        // Generated post quantum keys
        return kyberPreKeyRecords;
    }

    public KyberPreKeyRecord generateKyberPreKey(ECPrivateKey ecPrivateKey) {
        int kyberSignedPreKeyId = new SecureRandom().nextInt(Medium.MAX_VALUE - 1);
        KEMKeyPair kemKeyPair = KEMKeyPair.generate(KEMKeyType.KYBER_1024);
        KyberPreKeyRecord kyberPreKeyRecord = new KyberPreKeyRecord(kyberSignedPreKeyId,
                System.currentTimeMillis(), kemKeyPair,
                ecPrivateKey.calculateSignature(kemKeyPair.getPublicKey().serialize()));
        return kyberPreKeyRecord;
    }

    public void processKeyDistributionMessage(String name, String base64Key) {
        try {
            Log.i(TAG, "base64 keyDistributionMessage: " + base64Key);
            SenderKeyDistributionMessage senderKeyDistributionMessage =
                    new SenderKeyDistributionMessage(Base64.getDecoder().decode(base64Key));
            new GroupSessionBuilder(sekretessSignalProtocolStore)
                    .process(new SignalProtocolAddress(name, 1), senderKeyDistributionMessage);
            Log.i(TAG, "Group chat chipper created and stored : " + name);
        } catch (Exception e) {
            Log.e(TAG, "Error during decrypt key distribution message", e);
            if (apiClientBridge != null) {
                apiClientBridge.showToast("Error during decrypt distribution message: " + e.getMessage());
            }
        }
    }

    public void storeKyberPreKeyRecords(KyberPreKeyRecord[] kyberPreKeyRecords) {
        for (KyberPreKeyRecord kyberPreKeyRecord : kyberPreKeyRecords) {
            sekretessSignalProtocolStore.storeKyberPreKey(kyberPreKeyRecord.getId(), kyberPreKeyRecord);
            Log.i(TAG, "Storing KyberPreKEyRecord:" +kyberPreKeyRecord.getId());
        }
    }

    public void storePreKeyRecords(PreKeyRecord[] preKeyRecords) {
        for (PreKeyRecord preKeyRecord : preKeyRecords) {
            Log.i(TAG, "Storing PreKeyRecord:" +preKeyRecord.getId());
            sekretessSignalProtocolStore.storePreKey(preKeyRecord.getId(), preKeyRecord);
        }
    }

    public void storeSignedPreKey(SignedPreKeyRecord signedPreKeyRecord) {
        sekretessSignalProtocolStore.storeSignedPreKey(signedPreKeyRecord.getId(), signedPreKeyRecord);
    }

    public SignedPreKeyRecord generateSignedPreKey(ECKeyPair keyPair, byte[] signature) {
        //Generate signed prekeyRecord
        int signedPreKeyId = new SecureRandom().nextInt(Medium.MAX_VALUE - 1);
        return new SignedPreKeyRecord(signedPreKeyId, System.currentTimeMillis(), keyPair, signature);
    }

    public PreKeyRecord[] generatePreKeys() {
        PreKeyRecord[] preKeyRecords = new PreKeyRecord[SIGNAL_KEY_COUNT];
        SecureRandom preKeyRecordIdGenerator = new SecureRandom();
        for (int i = 0; i < preKeyRecords.length; i++) {
            int id = preKeyRecordIdGenerator.nextInt(Integer.MAX_VALUE);
            ECKeyPair ecKeyPair = ECKeyPair.generate();
            PreKeyRecord preKeyRecord = new PreKeyRecord(id, ecKeyPair);
            preKeyRecords[i] = preKeyRecord;
        }
        return preKeyRecords;
    }

    /**
     * Clears all Signal protocol keys from local storage (e.g. on logout).
     * Call this in parallel with removing auth state when the user logs out.
     */
    public void clearSignalKeys() {
        sekretessSignalProtocolStore.clearStorage();
    }

    public KeyBundle initializeKeyBundle() {
        sekretessSignalProtocolStore.clearStorage();

        ECKeyPair signedPreKeyPair = ECKeyPair.generate();
        IdentityKeyPair identityKeyPair = sekretessSignalProtocolStore.getIdentityKeyPair();
        int registrationId = sekretessSignalProtocolStore.getLocalRegistrationId();

        byte[] signature = identityKeyPair.getPrivateKey().calculateSignature(signedPreKeyPair
                .getPublicKey().serialize());

        //Generate one-time prekeys
        PreKeyRecord[] opk = generatePreKeys();

        SignedPreKeyRecord signedPreKeyRecord = generateSignedPreKey(signedPreKeyPair, signature);
        KyberPreKeyRecord[] kyberPreKeyRecords = generateKyberPreKeys(identityKeyPair.getPrivateKey());

        return new KeyBundle(registrationId, opk, signedPreKeyRecord,
                identityKeyPair, signature,
                kyberPreKeyRecords);
    }


    public boolean init() throws Exception {
        if (sekretessSignalProtocolStore.registrationRequired()) {
            KeyBundle keyBundle = initializeKeyBundle();

            if (apiClientBridge != null && apiClientBridge.upsertKeyStore(keyBundle)) {
                storeKyberPreKeyRecords(keyBundle.getKyberPreKeyRecords());
                storePreKeyRecords(keyBundle.getOpk());
                storeSignedPreKey(keyBundle.getSignedPreKeyRecord());
                return true;
            } else {
                Log.w(TAG, "Upsert cryptographic keys failed");
                sekretessSignalProtocolStore.clearStorage();
                return false;
            }
        } else if (sekretessSignalProtocolStore.updateKeysRequired()) {
            Log.w(TAG, "Update onetime cryptographic keys");
            updateOneTimeKeys();
        }
        return true;
    }

    public Optional<String> decryptGroupChatMessage(String sender, String base64Message) throws NoSessionException, InvalidMessageException, DuplicateMessageException, LegacyMessageException {
        GroupCipher groupCipher = new GroupCipher(sekretessSignalProtocolStore, new SignalProtocolAddress(sender, 1));
        String decryptedMessage = new String(groupCipher.decrypt(base64Decoder.decode(base64Message)));
        return Optional.of(decryptedMessage);
    }

    public Optional<String> decryptPrivateMessage(String sender, String base64Message) throws InvalidMessageException, InvalidVersionException, LegacyMessageException, InvalidKeyException, UntrustedIdentityException, DuplicateMessageException, InvalidKeyIdException {
        PreKeySignalMessage preKeySignalMessage = new PreKeySignalMessage(base64Decoder.decode(base64Message));
        SignalProtocolAddress signalProtocolAddress = new SignalProtocolAddress(sender, 1);
        SessionCipher sessionCipher = new SessionCipher(sekretessSignalProtocolStore, signalProtocolAddress);
        return Optional.of(new String(sessionCipher.decrypt(preKeySignalMessage, UsePqRatchet.YES)));
    }
}
